src/fast/u192.rs

```diff
@@ -0,0 +1,57 @@
+use std::mem::transmute;
+use std::ops::{Add, Mul};
+
+// This was started by the output of the uint crate,
+// then heavily reduced to only the parts we need
+// (which was a significant optimization, especially
+// in the mul which carried out the full multiplication
+// and discarded the top-half)
+//
+/// Little-endian large integer type
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+pub struct U192(pub [u64; 3]);
+
+impl Mul for U192 {
+    type Output = Self;
+    fn mul(self, other: Self) -> Self {
+        // The generated implementation of this method was 360 lines long!
+        let me = &self.0;
+        let you = &other.0;
+
+        let mult = |m: usize, y: usize| unsafe {
+            transmute::<_, (u64, u64)>(me[m] as u128 * you[y] as u128)
+        };
+
+        let (r0, r1) = mult(0, 0);
+        let (low, hi0) = mult(1, 0);
+        let (r1, overflow0) = low.overflowing_add(r1);
+        let (low, hi1) = mult(0, 1);
+        let (r1, overflow1) = low.overflowing_add(r1);
+
+        let r2 = (hi0 + overflow0 as u64)
+            .wrapping_add(hi1 + overflow1 as u64)
+            .wrapping_add(me[2].wrapping_mul(you[0]))
+            .wrapping_add(me[1].wrapping_mul(you[1]))
+            .wrapping_add(me[0].wrapping_mul(you[2]));
+
+        U192([r0, r1, r2])
+    }
+}
+
+impl Add for U192 {
+    type Output = Self;
+    fn add(self, other: Self) -> Self {
+        let me = &self.0;
+        let you = &other.0;
+
+        let (r0, overflow) = me[0].overflowing_add(you[0]);
+        let (res, overflow1) = me[1].overflowing_add(you[1]);
+        let (r1, overflow2) = res.overflowing_add(overflow as u64);
+
+        let r2 = me[2]
+            .wrapping_add(you[2])
+            .wrapping_add(overflow1 as u64 + overflow2 as u64);
```

Comment:
I don't understand the math here: AFAICT, we want to compute 
```
me[0] + you[0] + 2^64*(overflow + me[1] + you[1]) + 2^128*(me[2]+you[2] + overflow1)
```
IOW if we have an overflow from `me[0] + you[0]` it needs to be added to `me[1] + you[1]` and we have an overflow if `overflow2` is true.

Reply:
I think as-is is correct, but don't give up the fight unless you are totally convinced because it's important to get it right.

The reason we don't have an overflow if `overflow2` is true is because both `overflow1` and `overflow2` are tracking overflow of `r1`. One is from the possibility of overflow in the contribution of `me[1] + you[1]` to `r1` and one is from `r1 + overflow` to `r1`. These would be better named `overflow1a` and `overflow1b` and there is no `overflow2` since the contributions to `r2` use `wrapping_add`.

Reply:
I'm going to rename these to `overflow0`, `overflow1a` and `overflow1b`. If we did need to track overflow for `r2` it would be `overflow2a ... overflow2c`.

Reply:
You are right - this is correct; I was getting confused about the different `overflow` variables